core/chains/evm/headtracker/heads.go

```diff
@@ -26,114 +27,158 @@ type Heads interface {
 }
 
 type heads struct {
-	heads    []*evmtypes.Head
-	headsMap map[common.Hash]*evmtypes.Head
-	mu       sync.RWMutex
+	highest       *evmtypes.Head
+	headsAsc      *headsHeap
+	headsByHash   map[common.Hash]*evmtypes.Head
+	headsByParent map[common.Hash]map[common.Hash]*evmtypes.Head
+	mu            sync.RWMutex
 }
 
 func NewHeads() Heads {
-	return &heads{}
+	return &heads{
+		headsAsc:      &headsHeap{},
+		headsByHash:   make(map[common.Hash]*evmtypes.Head),
+		headsByParent: map[common.Hash]map[common.Hash]*evmtypes.Head{},
+	}
 }
 
 func (h *heads) LatestHead() *evmtypes.Head {
 	h.mu.RLock()
 	defer h.mu.RUnlock()
 
-	if len(h.heads) == 0 {
-		return nil
-	}
-	return h.heads[0]
+	return h.highest
 }
 
 func (h *heads) HeadByHash(hash common.Hash) *evmtypes.Head {
 	h.mu.RLock()
 	defer h.mu.RUnlock()
 
-	if h.headsMap == nil {
+	if h.headsByHash == nil {
 		return nil
 	}
 
-	return h.headsMap[hash]
+	return h.headsByHash[hash]
 }
 
 func (h *heads) Count() int {
 	h.mu.RLock()
 	defer h.mu.RUnlock()
 
-	return len(h.heads)
+	return h.headsAsc.Len()
 }
 
-// MarkFinalized - marks block with has equal to finalized and all it's direct ancestors as finalized.
+// MarkFinalized - marks block with hash equal to finalized and all it's direct ancestors as finalized.
 // Trims old blocks whose height is smaller than minBlockToKeep
 func (h *heads) MarkFinalized(finalized common.Hash, minBlockToKeep int64) bool {
 	h.mu.Lock()
 	defer h.mu.Unlock()
 
-	if len(h.heads) == 0 {
+	if len(h.headsByHash) == 0 {
 		return false
 	}
 
-	// deep copy to avoid race on head.Parent
-	h.heads, h.headsMap = deepCopy(h.heads, minBlockToKeep)
-
-	finalizedHead, ok := h.headsMap[finalized]
+	finalizedHead, ok := h.headsByHash[finalized]
 	if !ok {
 		return false
 	}
-	for finalizedHead != nil {
-		finalizedHead.IsFinalized = true
-		finalizedHead = finalizedHead.Parent
+
+	markFinalized(finalizedHead)
+
+	// remove all blocks that are older than minBlockToKeep
+	for h.headsAsc.Len() > 0 && h.headsAsc.Peek().Number < minBlockToKeep {
+		oldBlock := heap.Pop(h.headsAsc).(*evmtypes.Head)
+		delete(h.headsByHash, oldBlock.Hash)
```

Comment:
Would it be possible to delete and return a copy within `heap.Pop` instead?

Reply:
Not sure that I understand. Could you elaborate?

Reply:
I thought you were dereferencing the popped item from the Heap here, which actually leads me to the actual question I had: should we do this `h.values[len(h.values)-1] = nil` inside Pop to avoid memory leaks?